### ISSUE TITLE
sql: support scalar user-defined functions

### ISSUE THREAD

**Issue by jordanlewis** [cockroachdb/cockroach#58356]
This issue tracks adding basic user-defined scalar SQL function support. A scalar user-defined function is a user-defined function that only uses scalar projections in a `SELECT`.

For example,

```
create function f (a int, b int) returns int as 'select a+b' language sql;
```

There are quite a few modifiers that can be specified on `CREATE FUNCTION` according to the PostgreSQL documentation: https://www.postgresql.org/docs/13/sql-createfunction.html

Checklist for things that we'll want to make sure of by the time this is finished:

- [x] CREATE FUNCTION
- [x] CREATE OR REPLACE FUNCTION
- [x] DROP FUNCTION
- [x] UDFs should be present in pg_proc and other such metadata tables
- [ ] UDFs in views should get their dependencies tracked properly
- [ ] cached plans need to notice when UDFs are modified
- [ ] UDFs in computed columns should get their dependencies tracked properly
- [ ] Make sure that user-defined functions with custom types are going to be okay
- [x] Make sure that ::regproc casts work properly both to and from oid/name. See #61211.

Additionally, it turns out that to properly do name resolution for user-defined functions with overloads, there is significant extra complexity. We can't simply store each overload in the namespace table with its type parameters as part of the name, because that would require an uncachable range scan on every resolution to a given user-defined function.

So, what we need to do is to implement a 2-level descriptor tree for user-defined functions: the parent contains the name of the UDF, and the children contain the overload types and definition (and privileges, dependencies, etc). This is fairly irritating because leasing a user-defined function now requires pulling in all of the children overloads, and leasing them as well.

We'd like to follow the pattern that was used for leasing tables that contain user-defined types, which have a similar problem. The problem is solved with a special cache of user-defined types that is implemented in the `pkg/sql/catalog/hydratedtables` package.

Vague todo list:

- [x] flesh out the func overload descriptor protobuf
- [ ] add logic to create both the parent and child upon `CREATE FUNCTION`
- [x] generalize the `hydratedtables` package to also work for func overload descriptors
- [x] thread a func overload descriptor cache into the table collection to cache the func descriptors upon retrieval of the parent function descriptor

Jira issue: CRDB-11401

**Comment by rafiss** [comment]
I was just thinking of how there are many functions in pg_builtins.go that are implemented using the internal executor. This can get expensive (as in #57924) because the internal executor does not have the same descriptor collection as the external context. It would be cool if we could implement builtins using this!

**Comment by ajwerner** [comment]
> because the internal executor does not have the same descriptor collection as the external context. It would be cool if we could implement builtins using this!

If this is the root cause of some slowness, we should address this. It does not feel very hard.

**Comment by rafiss** [comment]
I agree, I'd want the UDF to be inlined as well.

I wasn't suggesting to use the internal executor for this. I was saying that implementing some builtins using the internal executor can be slow, so perhaps it could make more sense to implement those builtins using UDFs.

**Comment by mari-crl** [comment]
:information_source: Hello! I am a human and not at all a robot! Look at my very human username! :robot: :notes:
:thinking: Although I tried very hard to figure out what to do with this issue, more powerful human brains will need to help me. (specifically: Both Github and Jira issues were modified after exalate problems) :confounded:
:arrows_counterclockwise: **Please visit this issue's mirror at [CRDB-11401](https://cockroachlabs.atlassian.net/browse/CRDB-11401) and try to sync the two sides up manually.** :star2:
:white_check_mark: When you're finished, comment saying as much asn a member of Developer Infrastructure will be along to finish linking. :link:
:no_entry_sign: Note that until this is done, **this issue is not and will not be synced to Jira with Exalate**. :no_entry_sign:
:sweat_smile: Feeling lost? Don't worry about it! A member of @cockroachdb/exalate-22-cleanup-team will be along shortly to help! :+1:
:construction_worker: Developer Infrastructure members: when ready, open Exalate from the right-hand menu of the mirror issue in Jira, then choose Connect and enter this issue's URN: `cockroachdb/cockroach-58356`. Either way, delete this comment when you're done. :key:
:pray: Thank you for your compliance, my fellow humans! :robot: :wave:

**Comment by mgartner** [comment]
Scalar UDFs will be available in 22.2, so I'm going to close this issue. Note that not all types of UDFs will be supported in 22.2 - please see other UDF issues to track progress on missing features.
